### PR TITLE
refactor: changing the repo to be based on prior work established

### PR DESCRIPTION
**Pull request by ryoari**
### Refactors into a new baseline for successive work.

Due to collaborative effort of previous work, a lot of technical debt had built up where I wasn't sure of the code and it's effect on determinism.

As such, as I parse through the preexisting code to gain a better understanding. It is better to move everything to a legacy folder and re-introduce slowly after validating the code functions and it's effect.

Based on my prior supporting work in the https://github.com/ryoari/Verifiable-LLM-Baseline/

I rebase this repo to my previous effort, with clearly understood foundations.

### this PR builds on #73 #37 #86 

Adds in the following files:

## `src/`

| File | Summary |
|------|---------|
| `config.py` | Immutable training config + SHA-256 config hash |
| `dataset.py` | Tiny deterministic char-level dataset + batches |
| `device.py` | Device select (CPU/CUDA/XPU) + determinism setup |
| `eval.py` | Held-out eval, perplexity, sealed eval manifest |
| `global_manifest.py` | Combines all hashes into sealed pipeline manifest |
| `gpu_reproducibility_test.py` | Fresh-vs-fresh bitwise determinism check per device |
| `main.py` | `set_seed` entropy control + sanity smoke test |
| `model.py` | Deterministic NanoGPT (attention, block, TinyGPT) |
| `reproducibility.py` | Segmented prover/auditor + 5 tamper scenarios |
| `telemetry.py` | Per-step loss/grad/param logger + model hashing |

## `proofs/`

| File | Summary |
|------|---------|
| `cpu_determinism_log.txt` | Narrated CPU run history / verification log |
| `device_determinism_log.txt` | Machine-readable fresh-vs-fresh proof records |
| `env_fingerprint.json` | Torch/Python/OS/CPU environment snapshot |
| `eval_manifest.json` | Eval loss, perplexity, dataset + checkpoint hashes |
| `pipeline_manifest.json` | Final sealed global pipeline hash bundle |

### AI Usage Disclosure:
- [ ] This PR does not contain AI-generated code at all.
- [x] This PR contains AI-generated code. I have read the [AI Usage Policy](https://github.com/AOSSIE-Org/Info/blob/main/AI-UsagePolicy.md) and this PR complies with this policy. I have tested the code locally and I am responsible for it.

I have used the following AI models and tools: 
CLAUDE for comments and reasoning in device.py and gpu_reproducibility.py

## Checklist

- [x] My PR addresses a single issue, fixes a single bug or makes a single improvement.
- [x] My code follows the project's code style and conventions
- [x] If applicable, I have made corresponding changes or additions to the documentation
- [x] If applicable, I have made corresponding changes or additions to tests
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] Once I submit my PR, CodeRabbit AI will automatically review it and I will address CodeRabbit's comments.
- [x] I have filled this PR template completely and carefully, and I understand that my PR may be closed without review otherwise.

## Summary by CodeRabbit

* **New Features**
  * End-to-end determinism & reproducibility tooling: deterministic model + tiny dataset demo, automated fresh‑vs‑fresh bitwise checks, sealed evaluation and pipeline manifests, and a CLI entrypoint for deterministic setup and quick demos.

* **Documentation**
  * Added a Colab workflow demonstrating GPU reproducibility steps and interpretation.
  * Included proof logs and environment/evaluation manifests capturing reproducibility outcomes.

* **Chores**
  * Added dependency manifest and project ignores for local/dev artifacts.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+torch>=2.10.0,<3.0
+numpy>=2.4,<3.0
+tqdm>=4.67,<5.0
```
[the remark this answers is not in the document]
Comment (ryoari) [reply requirements.txt]: ignore for now, will fix to a particular dependency later

File: proofs/device_determinism_log.txt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  "torch": "2.12.0+cpu",
+  "accelerator_version": null,
+  "os": "Windows-11-10.0.22621-SP0",
+  "cublas_workspace_config": ":4096:8",
+  "final_loss": 0.2107149064540863,
+  "loss_curves_identical": true,
+  "model_hash_run1": "0fe9e61625b6c26825256b805aa6c862162df654644609e2e96e2efc672e473d",
+  "model_hash_run2": "0fe9e61625b6c26825256b805aa6c862162df654644609e2e96e2efc672e473d",
+  "bitwise_reproducible": true
+}
```
[the remark this answers is not in the document]
Comment (ryoari) [reply proofs/device_determinism_log.txt:52]: ignore, is needed for non-cuda

File: src/eval.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    dataset = TinyDataset()
+    model = TinyGPT(
+        vocab_size=dataset.vocab_size,
+        embed_dim=TRAIN_CONFIG["embed_dim"],
+        num_heads=TRAIN_CONFIG["num_heads"],
+        max_seq_len=TRAIN_CONFIG["max_seq_len"],
+        dropout=TRAIN_CONFIG["dropout"]
+        ).to(DEVICE)
+
+    checkpoint = torch.load("mid_checkpoint.pt", weights_only=False, map_location=DEVICE)
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/eval.py]: will check this out later

File: src/global_manifest.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    # 2. Configuration Hash
+    config_hash = get_config_hash()
+
+    # 3. Dataset Hash
+    dataset = TinyDataset()
+    dataset_hash = hashlib.sha256(dataset.encoded.numpy().tobytes()).hexdigest()
+
+    # 4. Model Hash 
+    with open("mid_checkpoint.pt", "rb") as f:
+        model_hash = hashlib.sha256(f.read()).hexdigest()
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/global_manifest.py]: will check this later

File: src/reproducibility.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        max_seq_len=TRAIN_CONFIG["max_seq_len"],
+        dropout=TRAIN_CONFIG["dropout"]
+        ).to(DEVICE)
+
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=TRAIN_CONFIG["lr"])
+    logger = TelemetryLogger(filepath=log_file)
+
+    if checkpoint_path_to_load:
+        checkpoint = torch.load(checkpoint_path_to_load, weights_only=False)
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/reproducibility.py:56]: will check this later

File: src/reproducibility.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        model.load_state_dict(checkpoint['model'])
+
+        # verifying cryptographic seal
+        if 'checkpoint_hash' in checkpoint:
+            loaded_hash = logger.hash_model(model)
+            if loaded_hash != checkpoint['checkpoint_hash']:
+                print(f"\n  FATAL ALERT! : Cryptographic seal broken! Checkpoint file was tampered with.")
+                print(f"    Expected: {checkpoint['checkpoint_hash'][:16]}...")
+                print(f"    Got:      {loaded_hash[:16]}...\n")
+
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/reproducibility.py:67]: it's a failure mode, should fail

File: src/reproducibility.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+    if checkpoint_path_to_load:
+        checkpoint = torch.load(checkpoint_path_to_load, weights_only=False)
+        model.load_state_dict(checkpoint['model'])
+
+        # verifying cryptographic seal
+        if 'checkpoint_hash' in checkpoint:
+            loaded_hash = logger.hash_model(model)
+            if loaded_hash != checkpoint['checkpoint_hash']:
+                print(f"\n  FATAL ALERT! : Cryptographic seal broken! Checkpoint file was tampered with.")
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/reproducibility.py]: later

File: src/reproducibility.py [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+
+    dataset = TinyDataset()
+    model = TinyGPT(vocab_size=dataset.vocab_size, embed_dim=TRAIN_CONFIG["embed_dim"], num_heads=TRAIN_CONFIG["num_heads"], max_seq_len=TRAIN_CONFIG["max_seq_len"], dropout=TRAIN_CONFIG["dropout"]).to(DEVICE)
+    optimizer = torch.optim.Adam(model.parameters(), lr=TRAIN_CONFIG["lr"])
+    logger = TelemetryLogger(filepath=log_file)
+
+    checkpoint = torch.load("mid_checkpoint.pt", weights_only=False)
+    model.load_state_dict(checkpoint['model'])
+    optimizer.load_state_dict(checkpoint['optimizer'])
+
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/reproducibility.py:158]: will check this later

File: src/reproducibility.py [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    hash_match = (prover_hash == auditor_hash)
+    if not hash_match:
+        print(f"\n Hash mismatch! Prover hash: {prover_hash[:16]} // Auditor hash: {auditor_hash[:16]} [HASH ERROR]")
+
+    if match and hash_match:
+        print(f"\n (❁ ´◡`❁) {label} PASSED: Segment replay is bitwise deterministic.")
+    else:
+        print(f"\n (╯°□°）╯︵ ┻━┻  {label} FAILED: Trajectories diverged.")
+
+    return match
```
[the remark this answers is not in the document]
Comment (ryoari) [reply src/reproducibility.py]: will check this later
